### PR TITLE
Cpp-Core guidelines and Modernization

### PR DESCRIPTION
**Pull request by revol-xut**
# This is not ready to merge yet

## Goals
- C++17 features all the way
- consistent use of space vs tabs now its all tabs
- consistent member variable naming convention was a mix of Hungarian notation and prefix dash. 
- making everything explicit (noexcept, const, inline) where possible
- changing the memory barriers of some atomics inside the scheduler

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -1,6 +1,202 @@
----
-BasedOnStyle: Chromium
-BreakConstructorInitializersBeforeComma: 'true'
-AllowShortFunctionsOnASingleLine: All
+Checks: 'clang-diagnostic-*,clang-analyzer-*,boost-*,Google-*,clang-analyzer-*,modernize-*,performance-*,portability-*,readability-*,cppcoreguidelines-*,llvm-*,cert-*'
+WarningsAsErrors: true
+HeaderFilterRegex: ''
+AnalyzeTemporaryDtors: false
+FormatStyle: none
```
Comment (cmnrd) [inline .clang-format:5]: Does this mean clang-format does not enforce any style now?

File: doc/reactor-cpp.doxyfile [hunk 2]
```diff
@@ -368,7 +368,7 @@ CPP_CLI_SUPPORT        = NO
 # Set the SIP_SUPPORT tag to YES if your project consists of sip (see:
 # https://www.riverbankcomputing.com/software/sip/intro) sources only. Doxygen
 # will parse them like normal C++ but will assume all classes use public instead
-# of private inheritance when no explicit protection keyword is present.
+# of private inheritance when no explicit protection keyword is present_.
```
Comment (cmnrd) [inline doc/reactor-cpp.doxyfile:371]: Why the underscore?

File: examples/count/main.cc [hunk 3]
```diff
@@ -11,12 +11,12 @@ class Count : public Reactor {
   Timer timer{"timer", this};
   LogicalAction<int> counter{"counter", this};
 
-  // reactions
+  // reactions_
```
Comment (cmnrd) [inline examples/count/main.cc:14]: Also here, why is the underscore added?

File: include/reactor-cpp/action.hh [hunk 4]
```diff
@@ -15,159 +15,175 @@
...
-      , min_delay(min_delay) {}
+        : ReactorElement(name, ReactorElement::Type::Action, container)
+        , logical_(logical)
+        , min_delay(min_delay) {}
 
- public:
-  const auto& triggers() const { return _triggers; }
-  const auto& schedulers() const { return _schedulers; }
+public:
+    [[nodiscard]] auto triggers() const noexcept -> const auto& {
```
Comment (cmnrd) [inline include/reactor-cpp/action.hh:41]: Why does this change to the `->` notation? Is this recommended by the guidelines?

**Review by cmnrd — COMMENTED** [review]
The changes are massive, which makes this hard to review. For now, I had a quick glance at it and added a few comments, but I will do a more thorough review in the future.

**Comment by revol-xut** [comment]
@cmnrd Thanks for the feedback to be honest I am not really done yet so it's not really worth it to review it just yet.

File: include/reactor-cpp/action.hh (hunk 4 above)
Comment (revol-xut) [reply include/reactor-cpp/action.hh]: no I don't think so but it's included inside the clang-tidy modernization check.

**Comment by revol-xut** [comment]
@cmnrd very interesting. Github Actions uses your newly defined actions as well as mine. I am very surprised

**Comment by revol-xut** [comment]
Sure

File: lib/reactor.cc [hunk 5]
```diff
@@ -34,120 +33,132 @@ ReactorElement::ReactorElement(const std::string& name,
...
-  toggle_assert(action != nullptr);
-  reactor::validate(this->environment()->phase() == Environment::Phase::Construction,
-           "Actions can only be registered during construction phase!");
-  toggle_assert(_actions.insert(action).second);
+void Reactor::register_action([[maybe_unused]] BaseAction *action) {
+  reactor_assert(action != nullptr);
+  reactor::validate(
+      this->environment()->phase() == Environment::Phase::Construction,
+      "Actions can only be registered during construction phase!");
+  reactor_assert(actions_.insert(action).second);
```
Comment (cmnrd) [inline lib/reactor.cc:80]: I noticed a problem in the `register_*` methods while working on a bugfix on master. Why is the method argument labeled as `[[maybe_unused]]`? It should always be used and inserted into the according set.
I suspect that if assertions are off, the insertion is not executed. I wonder if this was different when we used the C style assert... Anyway, I think the right way of implementing this would be to always do the insertion and only assert the result.

File: include/reactor-cpp/port.hh [hunk 6]
```diff
@@ -18,122 +19,112 @@ namespace reactor {
...
- private:
-  bool present{false};
-
-  void cleanup() override final { present = false; }
+template <> class Port<void> : public BasePort {
+private:
+  bool present_ = false;
 
- public:
-  using value_type = void;
```
Comment (cmnrd) [inline include/reactor-cpp/port.hh:103]: `value_type` should be set in both the general and the specialized version of `Port`

File: include/reactor-cpp/port.hh [hunk 7]
```diff
@@ -18,122 +19,112 @@ namespace reactor {
...
 
-template <>
-class Port<void> : public BasePort {
- private:
-  bool present{false};
-
-  void cleanup() override final { present = false; }
+template <> class Port<void> : public BasePort {
+private:
+  bool present_ = false;
```
Comment (cmnrd) [inline include/reactor-cpp/port.hh:93]: Why did this change from brace initialization to equal initialization? As far as I understand, there are good reasons to use brace initialization wherever possible. See [this discussion](https://stackoverflow.com/questions/18222926/why-is-list-initialization-using-curly-braces-better-than-the-alternatives) for instance.

File: include/reactor-cpp/port.hh [hunk 8]
```diff
@@ -18,122 +19,112 @@ namespace reactor {
...
-template <class T>
-class Input : public Port<T> {
- public:
+template <class T> class Input : public Port<T> { // NOLINT
+public:
   Input(const std::string& name, Reactor* container)
       : Port<T>(name, PortType::Input, container) {}
-
-  Input(Input&&) = default;
+  Input(Input&&) noexcept = default;
```
Comment (cmnrd) [inline include/reactor-cpp/port.hh:116]: Could it be that this change triggers the build problems with gcc 9.3? For some reason, the compiler rejects the Input type as not move-assignable. This change here modifies the constructor signature, and it might be the reason why the constructor is not detected as a move-constructor.

**Comment by cmnrd** [comment]
Looks like [44cd7dd](https://github.com/lf-lang/reactor-cpp/pull/6/commits/44cd7ddfc135de19a076944c82d4fc6c87d39b4f) indeed fixed the compile errors, right?

I see that there are validation errors thrown. Those likely indicate problems in the existing validation code (as apparently validations haven't worked in the past...)

**Comment by revol-xut** [comment]
@cmnrd yes the removal of the `noexcept` fixed the strange compilation bug we got.

File: examples/count/main.cc [hunk 9]
```diff
@@ -6,43 +6,43 @@ using namespace reactor;
...
-    auto& value = *(counter.get());
-    std::cout << "Count: " << value << std::endl;
-    counter.schedule(value + 1, 1s);
-  }
-
-  void init() {
-    std::cout << "Hello World!" << std::endl;
-    counter.schedule(0, 1s);
-  }
+private:
```
Comment (cmnrd) [inline examples/count/main.cc:9]: doesn't look like this file has been formatted with the current .clang-forma (it used 4 spaces for indent, while 2 spaces are configured). Can we enforce correct formatting by running clang-format in CI and error out if it would do any reformatting?

File: flake.nix [hunk 10]
```diff
@@ -0,0 +1,29 @@
+{
+  description = "build script for the lingua-franca alarm clock";
```
Comment (cmnrd) [inline flake.nix:2]: The description does not fit ;)

File: include/reactor-cpp/trace.hh [hunk 11]
```diff
@@ -99,15 +64,15 @@ TRACEPOINT_EVENT(
...
+} // namespace reactor
 
 // empty definition in case we compile without tracing
 #define tracepoint(...)
 
-#endif  // _REACTOR_CPP_TRACE_H
+#endif // _REACTOR_CPP_TRACE_H
 
-#endif // REACTOR_CPP_TRACE
+#endif // REACTOR_CPP_TRACE_HH
```
Comment (cmnrd) [inline include/reactor-cpp/trace.hh:78]: This shouldn't change

File: include/reactor-cpp/trace.hh [hunk 12]
```diff
@@ -99,15 +64,15 @@ TRACEPOINT_EVENT(
 #else
 
 #ifndef _REACTOR_CPP_TRACE_H
-#define _REACTOR_CPP_TRACE_H
```
Comment (cmnrd) [inline include/reactor-cpp/trace.hh:102]: This shouldn't change (or `_REACTOR_CPP_TRACE_H` needs to be renamed everywhere)

**Review by cmnrd — COMMENTED** [review]
I made another partial review pass and added a few comments.

File: include/reactor-cpp/assert.hh [hunk 13]
```diff
@@ -20,32 +21,98 @@ constexpr bool runtime_assertion = false;
 constexpr bool runtime_assertion = true;
 #endif
 
+#ifdef __linux__
+constexpr bool linux_system = true;
+#else
+constexpr bool linux_system = true;
```
Comment (cmnrd) [inline include/reactor-cpp/assert.hh:27]: this seems wrong ;)

File: include/reactor-cpp/impl/port_impl.hh [hunk 14]
```diff
@@ -6,53 +6,55 @@
...
   // connected with another Port<T>.
   return static_cast<Port<T>*>(inward_binding());
 }
 
 template <class T>
 void Port<T>::set(const ImmutableValuePtr<T>& value_ptr) {
-  reactor::validate(!has_inward_binding(),
-           "set() may only be called on a ports that do not have an inward "
-           "binding!");
+  reactor::validate(!has_inward_binding(), "set() may only be called on a ports that do not have an inward "
```
Comment (cmnrd) [inline include/reactor-cpp/impl/port_impl.hh:34]:
```suggestion
  reactor::validate(!has_inward_binding(), "set() may only be called on ports that do not have an inward "
```

File: include/reactor-cpp/config.hh.in [hunk 15]
```diff
@@ -1,3 +1,3 @@
 #cmakedefine REACTOR_CPP_TRACE
-#cmakedefine REACTOR_CPP_VALIDATE
-#cmakedefine REACTOR_CPP_LOG_LEVEL @REACTOR_CPP_LOG_LEVEL@
+#cmakedefine REACTOR_CPP_VALIDATE false
```
Comment (cmnrd) [inline include/reactor-cpp/config.hh.in:2]: Why this change?

File: result [hunk 16]
```diff
@@ -0,0 +1 @@
+/nix/store/cyi4n90zglwd8d37487kjmkspy70whnx-all-tests
```
Comment (cmnrd) [inline result:1]: What is this?

File: tracing/README.md [hunk 17]
```diff
@@ -43,8 +43,8 @@ within `LD_LIBRARY_PATH`.
 The trace can be viewed by opening `about://tracing` in Chrome (or
 Chromium). There you can load the previously generated json file which produces
 a visualization of the trace. In this visualization,  the top part (labeled *Execution*),
-shows the physical time at which reactions execute. The bottom part shows individual reactors and logical times at which
-actions are scheduled and reactions are triggered.
+shows the physical time at which reactions_ execute. The bottom part shows individual reactors and logical times at which
```
Comment (cmnrd) [inline tracing/README.md:46]:
```suggestion
shows the physical time at which reactions execute. The bottom part shows individual reactors and logical times at which
```

File: tracing/README.md [hunk 18]
```diff
@@ -43,8 +43,8 @@ within `LD_LIBRARY_PATH`.
...
+actions are scheduled and reactions_ are triggered.
```
Comment (cmnrd) [inline tracing/README.md:47]:
```suggestion
actions are scheduled and reactions are triggered.
```

File: include/reactor-cpp/scheduler.hh [hunk 19]
```diff
@@ -27,123 +28,117 @@ namespace reactor {
...
 
   /**
-   * Fill the queue up with ready reactions.
+   * Fill the queue_ up with ready reactions_.
    *
-   * This method assumes that the internal queue is empty. It moves all
-   * reactions from the provided `ready_reactions` vector to the internal queue,
-   * leaving `ready_reactions` empty.
+   * This method assumes that the internal queue_ is empty. It moves all
+   * reactions_ from the provided `ready_reactions` vector to the internal
```
Comment (cmnrd) [inline include/reactor-cpp/scheduler.hh:78]: The renames shouldn't affect the comments in this file.

File: include/reactor-cpp/scheduler.hh [hunk 20]
```diff
@@ -27,123 +28,117 @@ namespace reactor {
...
 
-  Environment* _environment;
-  std::vector<Worker> workers{};
+  Environment* environment_;
+  std::vector<Worker> workers_{};
 
-  std::mutex m_schedule;
-  std::unique_lock<std::mutex> schedule_lock{m_schedule, std::defer_lock};
-  std::condition_variable cv_schedule;
+  std::mutex schedule_;
```
Comment (cmnrd) [inline include/reactor-cpp/scheduler.hh:97]: The name is misleading now. Rename to `lock_schedule_`? Or maybe better `schedule_mutex_`?

File: include/reactor-cpp/scheduler.hh [hunk 21]
```diff
@@ -27,123 +28,117 @@ namespace reactor {
...
 
-  ReadyQueue ready_queue;
-  std::atomic<std::ptrdiff_t> reactions_to_process{0};
+  ReadyQueue ready_queue_;
+  std::atomic<std::ptrdiff_t> reactions_to_process_{0}; //NOLINT
 
-  void schedule();
-  bool schedule_ready_reactions();
+  std::atomic<bool> stop_{false};
+  bool continue_execution_ ={true};
```
Comment (cmnrd) [inline include/reactor-cpp/scheduler.hh:114]:
```suggestion
  bool continue_execution_{true};
```

File: include/reactor-cpp/time.hh [hunk 22]
```diff
@@ -6,30 +6,31 @@
...
 
-#pragma once
+#ifndef REACTOR_CPP_TIME_HH
+#define REACTOR_CPP_TIME_HH
 
+#include "logging.hh"
 #include <chrono>
 #include <iostream>
 
+using TimePoint = std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>;
```
Comment (cmnrd) [inline include/reactor-cpp/time.hh:16]: Looks like there are two definitions of TImePoint and Duration

File: include/reactor-cpp/time.hh [hunk 23]
```diff
@@ -6,30 +6,31 @@
...
-std::ostream& operator<<(std::ostream& os, std::chrono::microseconds dur);
-std::ostream& operator<<(std::ostream& os, std::chrono::nanoseconds dur);
+auto operator<<(std::ostream &out_stream, reactor::TimePoint tp) noexcept -> std::ostream &;
+auto operator<<(std::ostream &out_stream, std::chrono::seconds dur) noexcept -> std::ostream &;
+auto operator<<(std::ostream &out_stream, std::chrono::milliseconds dur) noexcept -> std::ostream &;
+auto operator<<(std::ostream &out_stream, std::chrono::microseconds dur) noexcept -> std::ostream &;
+auto operator<<(std::ostream & out_stream, std::chrono::nanoseconds dur) noexcept -> std::ostream &;
 
-}  // namespace operators
+namespace reactor::operators {}
```
Comment (cmnrd) [inline include/reactor-cpp/time.hh:33]: What happened to the `operators` namespace? Why doesn't it contain the stream operators anymore?

**Review by cmnrd — COMMENTED** [review]
More comments from another partial review. I finished checking all the header files and still have the src files to go.

File: .github/workflows/clang-tidy.yml [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+      - name: Prepare compile_commands.json
+        run: |
+          cmake -B build -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
+          ls -a build
+      - name: Create results directory
+        run: |
+          mkdir clang-tidy-result
+      - name: Analyze
+        run: |
+          clang-tidy -p build lib/*.cc
```
[the remark this answers is not in the document]
Comment (revol-xut) [reply .github/workflows/clang-tidy.yml:24]: For archive purposed: the compile database that clang-tidy used only has entries for .cc files.

File: lib/environment.cc [hunk 25]
```diff
@@ -313,29 +164,49 @@ void Environment::export_dependency_graph(const std::string& path) {
   log::Info() << "Reaction graph was written to " << path;
 }
 
+void Environment::dump_to_yaml(const std::string &path) {
```
Comment (cmnrd) [inline lib/environment.cc:167]: Maybe move this down to the other `dump_*` methods?

File: lib/port.cc [hunk 26]
```diff
@@ -14,17 +14,15 @@
...
+  reactor_assert(this->environment() == port->environment()); //NOLINT
   validate(!port->has_inward_binding(), "Ports may only be connected once");
   validate(!this->has_dependencies(),
            "Ports with dependencies may not be connected to other ports");
   validate(!port->has_antidependencies(),
            "Ports with antidependencies may not be connected to other ports");
-  validate(this->environment()->phase() == Environment::Phase::Construction,
-           "Ports can only be bound during contruction phase!");
-
+  assert_phase(this, Environment::Phase::Assembly); //TODO:
```
Comment (cmnrd) [inline lib/port.cc:25]: Why the `//TODO`?

File: lib/reaction.cc [hunk 27]
```diff
@@ -6,6 +6,10 @@
  *   Christian Menard
  */
 
+#include <utility>
```
Comment (cmnrd) [inline lib/reaction.cc:9]: Duplicated include. Also, is this include needed? I didn't see any changes that would need the additional header.

File: include/reactor-cpp/port.hh (hunk 6 above)
Comment (cmnrd) [reply include/reactor-cpp/port.hh:105]: This was marked as resolved, but actually the diff still removes `value_type`.

File: include/reactor-cpp/scheduler.hh [hunk 28]
```diff
@@ -27,123 +28,117 @@ namespace reactor {
...
 class Worker;
 
-class Worker {
- public:
-  Scheduler& scheduler;
-  const unsigned id{0};
-  std::thread thread;
+class Worker { // NOLINT
+public:
+  Scheduler& scheduler_;
```
Comment (cmnrd) [inline include/reactor-cpp/scheduler.hh:33]: In the *.cc files, the `&` and `*` would be placed next to the identifier and not next to the type. Why is this different in the header files?

File: include/reactor-cpp/scheduler.hh [hunk 29]
```diff
@@ -27,123 +28,117 @@ namespace reactor {
...
   void execute_reaction(Reaction* reaction) const;
 
- public:
-  Worker(Scheduler& scheduler, unsigned id)
-      : scheduler{scheduler}, id{id}, thread{} {}
-  Worker(Worker&& w);
-  Worker(const Worker&) = delete;
+  Worker(Scheduler& scheduler, unsigned int identity)
+      : scheduler_{scheduler}
+      , identity_{identity} {}
```
Comment (cmnrd) [inline include/reactor-cpp/scheduler.hh:44]: Also this syntax is different in the *.cc files. While here it lists `,` under `:` (which personally I prefer), but it in the *.cc files it was formatted like this: `: scheduler_{scheduler}, identity_{identity} {}`

File: include/reactor-cpp/trace.hh [hunk 30]
```diff
@@ -66,56 +66,42 @@ TRACEPOINT_EVENT(
...
+    reactor_cpp, schedule_action,
+    TP_ARGS(const std::string&, reactor_name_arg, const std::string&, action_name_arg, const reactor::Tag&, tag_arg),
+    TP_FIELDS(ctf_string(reactor_name, reactor_name_arg.c_str()) ctf_string(action_name, action_name_arg.c_str())
+                  ctf_integer(unsigned long, timestamp_ns, tag_arg.time_point().time_since_epoch().count())
+                      ctf_integer(unsigned, timestamp_microstep, tag_arg.micro_step())))
+
+TRACEPOINT_EVENT(reactor_cpp, trigger_reaction,
+                 TP_ARGS(const std::string&, reactor_name_arg, const std::string&, reaction_name_arg,
+                         const reactor::LogicalTime&, tag_arg),
+                 TP_FIELDS(ctf_string(reactor_name, reactor_name_arg.c_str()) ctf_string(reaction_name,
```
Comment (cmnrd) [inline include/reactor-cpp/trace.hh:84]: Is there a way to disable auto formatting locally? Looks like clang-format does not play well with the TP_FIELDS macros.

File: lib/scheduler.cc [hunk 31]
```diff
@@ -55,306 +57,305 @@ void Worker::work() const {
...
-  auto old_size = size.exchange(new_size, std::memory_order_acq_rel);
+  auto new_size = static_cast<std::ptrdiff_t>(queue_.size());
+  auto old_size = size_.exchange(new_size, std::memory_order_acq_rel);
 
-  // calculate how many workers to wake up. -old_size indicates the number of
-  // workers who started waiting since the last update.
-  // We want to wake up at most all the waiting workers. If we would release
-  // more, other workers that are out of work would not block when acquiring
+  // calculate how many workers_ to wake up. -old_size indicates the number of
+  // workers_ who started waiting since the last update.
```
Comment (cmnrd) [inline lib/scheduler.cc:129]: names shouldn't change in the comments. (This goes for all the comments  in this file)

File: lib/scheduler.cc [hunk 32]
```diff
@@ -6,6 +6,8 @@
  *   Christian Menard
  */
 
+#include <utility>
```
Comment (cmnrd) [inline lib/scheduler.cc:9]: Is this include needed?

**Review by cmnrd — COMMENTED** [review]
Looks great! I added a few more comments.

File: include/reactor-cpp/scheduler.hh (hunk 29 above)
Comment (revol-xut) [reply include/reactor-cpp/scheduler.hh:44]: ok so u want it completely identically formatted okay. Understandable.

File: include/reactor-cpp/trace.hh (hunk 30 above)
Comment (revol-xut) [reply include/reactor-cpp/trace.hh:84]: not sure we have line length 120 right ?

**Review by cmnrd — APPROVED** [review]
Great! Thanks for all the hard work @revol-xut!
